New maze pathfinding

philgekni opened the pull request:
In the current maze pathfinding algorithm when a guest encounter a intersection, this one actually picks a random new direction. So, virtually more the maze contains intersections more difficult will be for the guests to get out, this independently of the complexity of the maze design.
Also it is a pretty common thing for guests to loop several time in the same area :

![sillyGuest.gif](https://user-images.githubusercontent.com/29202864/71312267-2b7b9580-2429-11ea-908b-b6527bf8d8ff.gif)

Considering such behaviors as visually silly (or even really dumb), giving a bad entrance throughput and presenting the risk for some guest to haunts forever moderately large mazes, I offer a new version of the maze pathfinding.

## New algorithm details
The principle of the new algorithm is based on path marking: at each intersection visitor will try to take a path that he doesn't already visited. A path is considered visited if all the dead-ends included into this path has been reached:

![mazeExploreAllPathsGraph](https://user-images.githubusercontent.com/29202864/71314015-ae104f00-2441-11ea-8a4c-63657a4b26d7.gif)

### Direction picking and possible other features
As say before guests will try to avoid choosing paths already visited: each time before picking a direction, a random check is run to determinate if the guest succeed to remember already visited path.

Each guest can remember the visited paths at the 5 last encountered intersections.

The probability to pick a wrong way is the result of a random value compared to a threshold value, which depends of how long time the guest has visited this intersection the last time.

If a guest jump / try to peek his position, at the next intersection he is more likely to choose an unvisited path. Also the guests have less probability to choose the first edge from where they entered in the intersections (to encourage them to visit all the ways in a direction before retracing their steps).

After some discussion on the dev Gitter, here some features / changes suggestions :

- [x] Randomly forget some visited paths by the guest. (@tupaschoal).
   _Implemented: guests remember the last 5 intersections and forget the others._
- [x] The probability to pick a wrong way also based on the time passed since the guest visited the intersection the last time. (@IntelOrca).
  _Implemented: The probabilities change in function of the position of the intersection into the guest memory of the last intersections_.
- *Open to new suggestions.*

## Implementation details
This version of the maze pathfinding is based on the feature that each guest remembers details concerning the 5 last intersections encountered on their way.

![mazeData](https://user-images.githubusercontent.com/29202864/72084567-054c6580-3304-11ea-8c18-26bcb996d776.png)

All the data used for this algorithm live inside the `Peep` structure in the `pathfind_history` fields (which are not used when the guest is in a ride) meaning that everything is saved into the save file and transmitted to new clients when joining a server.

## Results
![mazeResultV2:CurrentAlgorithm:204guestPerHour,NewAlgoritm:924guestPerHour](https://user-images.githubusercontent.com/29202864/72081811-70476d80-32ff-11ea-9330-26ea5944db37.png)
(_Results after one year from the same save file and with the weather locked to "Sunny"._) 

According to my first observations, I observed an large improvement of the throughput of the mazes. Also guests inside the maze are less complaining about their desire to get out of the maze:

![mazeThoughts](https://user-images.githubusercontent.com/29202864/72085418-8f48fe00-3305-11ea-9269-5813eeb0f257.png)

## Try by yourself
* [MazeTest.zip](https://github.com/OpenRCT2/OpenRCT2/files/5792795/MazeTest.zip): Include a save with some mazes I used to test this algorithm (new version 10/01/21).

## Roadmap
In side of the possibles review comments / changes and new ideas discussions here my agenda regarding this PR.

- [x] Plays with Peep pathfinding fields and write a new version of the algorithm.
- [x] Fix initial tests failures.
- [x] Coding style enforcement.
- [x] Add more tests / replace some `assert` to check in release possible errors situation due to hack done with cheating tools (like changing the height of a maze tile with the tile inspector).
- [ ] Tuning threshold values in regards of throughput and satisfaction results.

janisozaur commented:
There are failures highlighted in our CIs, please address them. Some compilation failures and some test suite failures.

philgekni commented:
Yup, on the way.

ZehMatt commented:
Heres what I think:
- For multiplayer when new clients join the list should not be cleared, players should not influence the game state like that.
- Storage can be probably simplified, the peep already has 4 slots for the path finding history, additionally you can also use pathfind_goal which makes a total of 5 junctions you could remember and allowing the data to be serialized in the save files. Having a total of 5 junctions in the peeps memory should already greatly improve some maze situations without having a perfect maze explorer.

philgekni commented:
I updated my branch with the last @tupaschoal `Map::map_get_track_element_at()` refactor and I took a look to the CIs.
All builds with tests fail due to MultiLaunch, result expected because for the moment I didn't yet implemented the data cleaning when the game is reloaded (see Roadmap). My only problem comes from one remaining compilation error on the Travis build job "Ubuntu i686 GCC Debug m32 no-build-shared-libs" (`undefined reference to 'map_get_track_element_at(CoordsXYZ const&)'`). As far I know I correctly include "world/Map.h" and didn't investigate yet too much.
@ZehMatt, thanks for the input and the suggestion about the pathfinding fields of the `Peep` structure (as I tried to avoid editing existing structures, I was too shy to dare to modify their values). I think that I will be indeed able to use these fields to save memory and to avoid inopportune data cleaning / regeneration in my code.
I will do some hack in this way in the next days and read a little the functions accessing to these values in order to understand what I risk to break.

janisozaur commented:
The link failure is in one of our special targets, `testpaint`. I think one of the team members can fix this for you.

philgekni commented:
I updated my branch with a new version of my algorithm with as key improvements:
- Complies with @ZehMatt concerns. Stop using extra memory which need to be wipe when reloading a game / a client joins the server. Everything lives inside the `Peep` structure and use `pathfind_history` memory. No memory allocation nor extra memory is needed.
- The new version is simpler. The intersection tracking is done by each guest for the last 5 encountered intersections. There is no more functions looping on maze tiles for data initialization or updating.
- I implemented in some ways the ideas that @tupaschoal and @IntelOrca gave me on Gitter (regarding paths and intersections forgetting).
- The guests are encouraged to discover new paths and not retracing their steps while they didn't completely visited the part of the maze where they are.
- The dead ends in form of loop are now detected and marked as visited after one loop.
- The debug panel of the guest window show the current state of the data used for the maze pathfinding.

I updated the PR description accordingly.

ZehMatt commented:
Looks pretty promising now except the coding style, please follow this guide regarding our code style: https://github.com/OpenRCT2/OpenRCT2/wiki/Coding-Style

I'll get to test and review your code shortly, probably this weekend.

pizza2004 commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4909,17 +4915,35 @@ void Guest::UpdateRideMazePathfinding()
+            { maze_set_odd(0.09), maze_set_odd(0.03), maze_set_odd(0.03), maze_set_odd(0.01) }, // -1
+            { maze_set_odd(0.27), maze_set_odd(0.07), maze_set_odd(0.18), maze_set_odd(0.10) }, // -2
+            { maze_set_odd(0.39), maze_set_odd(0.15), maze_set_odd(0.25), maze_set_odd(0.17) }, // -3
+            { maze_set_odd(0.50), maze_set_odd(0.31), maze_set_odd(0.45), maze_set_odd(0.32) }, // -4
```
Why is it that visiting is less effective than origin, but visited+peeked is more effective than origin+peeked a couple intersections back?  I would expect the relation between the two to be related, ie either you tend to forget your origin faster than your visited or peeking helps them both equally.

philgekni replied:
You right, the odds relations are her inconstant and I fixed that in 65573e6.
Regarding the fact that guests tends to forget less easily the origin than other visited directions, this is to encourage them to explore completely an area before turning back and so avoiding multiples travels between 2 area of the maze.

ZehMatt commented:
Sorry for the late response. I've tested this offline and online and seems to be working great, this improved the overall throughput for a few maze designs I tried out and had no issues. Before reviewing the code it would help if you could rebase and and change the coding style to match the base, overall pretty good job.

philgekni commented:
Thanks @ZehMatt, no worry for the delay, I have been myself occupied the last weeks. I'll find a moment this week to update my branch.

pizza2004 commented:
This will need to be rebased again before it can be merged, but I'm just wondering if anyone has done anything more with this in the past month?

philgekni commented:
Hi! No you had the right feeling, I haven't had time to work on this PR these last week but I'm now available.

I currently struggle to apply correctly the coding style. I believe that `clang-format` returns advises which are conflicting with the coding style, like:
```diff
 @@ -342,9 +342,8 @@
         item.free();
 }
 
-std::tuple<MazePathfindingEntry&, uint8_t, bool> MazePathfindingHistory::meetIntersection(const CoordsXY& coords,
-                                                                                          uint8_t subTileIndex,
-                                                                                          Direction source)
+std::tuple<MazePathfindingEntry&, uint8_t, bool> MazePathfindingHistory::meetIntersection(
+    const CoordsXY& coords, uint8_t subTileIndex, Direction source)
 {
     uint8_t i;
     bool hasPeeked = Stack[0].hasPeeked();
```
with from the [coding-style doc](https://github.com/OpenRCT2/OpenRCT2/wiki/Coding-Style):
```c
/// If a function definition runs over the line length limit, the arguments
/// should be broken up as follows:
void ShowAnimation(sint32 columns,
                   sint32 rows,
                   sint32 isResizable)
{
```

Let me know if I miss something.

pizza2004 commented:
If @philgekni no longer has time to work on this I might be interested in cleaning up the formatting and such so it could be merged and then any other improvements could be submitted later as new PRs.

philgekni commented:
Hi! I think that my PR is enough mature after the last refactorings and cleanings.
I tested the network locally and didn't encountered desync and also updated the mazeTest save (in the description) with new 'testcases'. If needed I can join a new version of the bpb1000Ticks replay taking in account the new guests behavior.

tupaschoal commented:
Nice. We'll probably be releasing a version this weekend, so I marked it as after 0.3.0, give us a couple of days

duncanspumpkin commented on src/openrct2/ride/gentle/Maze.h:
```diff
@@ -0,0 +1,180 @@
+    {
+        struct
+        {
+            uint32_t SubTileIdx : 2;       // Index of the subTile corresponding to the intersection
```
Do not use bit masks in the code. It is not multi-platform friendly. You will find that we never use them across the whole codebase for this reason. Instead we have lots of getters and setters. I'm not a fan of stuffing all this information into the struct like this. I would much rather we wait for the new save format and used individual variables of a sensible size.

pizza2004 replied:
If it's gonna be much longer before the new save file, it might be nice to just find a cross platform way to make it work in the meantime for the sake of allowing it to be merged.

philgekni replied:
Yeah, indeed, the bitfields are implementation dependent, my bad. So I think for the moment reverting aeba1ff and continuing to use bitwise ops, shifts and masks until the new save format. That should fixes the trouble regarding on how the different compilers want to organize the memory and lets only the endianness which is handled by the serialization module.

ZehMatt commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+    YTile = coords.y / COORDS_XY_STEP;
+}
+
+std::tuple<CoordsXY, uint8_t> MazePathfindingEntry::getCoords() const
```
Use std::pair for pairs

IntelOrca replied:
What is the benefit of using pair of tuple?

ZehMatt replied:
The compiler can optimize std::pair knowing it has two elements potentially into a single register, this doesn't seem to work with tuples.

IntelOrca replied:
The optimisation problem is a shame. Otherwise `std::pair` would be obsolete since `std::tuple` does the same thing but also allows more things. `std::get<N>` is the way I get items from both pair and tuple.

ZehMatt replied:
I find tuples to be generally a bad concept, just use a struct if you have more than 2 things. Having std::get<0>(var) is not exactly great code to read and one can only guess what the field could be.

duncanspumpkin replied:
You shouldn't need to use get<...> Very often when you could use structured bindings.
```
auto [var1, var2] = functionReturningTupleOrStructOrPair();```

ZehMatt replied:
This is true but that still means the caller needs to be aware of whats returned, all solved by structs.

philgekni replied:
Fixed by ebb203f.

ZehMatt commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+        item.free();
+}
+
+std::tuple<MazePathfindingEntry&, uint8_t> MazePathfindingHistory::meetIntersection(
```
Use std::pair for pairs

philgekni replied:
Fixed by ebb203f.

ZehMatt commented:
We generally use PascalCase for the function names

ZehMatt commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+
+#pragma region Maze Pathfinding
+
+void swap(MazePathfindingEntry& a, MazePathfindingEntry& b)
```
I don't think this function should exist

philgekni replied:
Fixed by Fixed by 1faeb39.

philgekni commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
     targetLoc.y += DestinationY;
 
-    enum class maze_type
+    enum class maze_next_tile_type
```
This enum is used locally (so that why I permitted me this modification) and I changed its name to reflect better that the code searches to characterize the type of the next tile and not the type of a maze in its entirety.

philgekni commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+
+#pragma region Maze Pathfinding
+
+void MazePathfindingEntry::Swap(MazePathfindingEntry& entry)
```
Right, I don't know why I was so attached to this function. Fixed by 1faeb39.

philgekni commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+{
+    uint8_t i;
+
+    for (i = 0; i < 5; ++i)
```
Fixed by 1faeb39.

philgekni commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+    for (int j = std::min<int>(4, i); j > 0; --j)
+        Stack[j - 1].Swap(Stack[j]);
+
+    if (i == 5)
```
Fixed by 1faeb39.

philgekni commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+            break;
+    }
+
+    for (int j = std::min<int>(4, i); j > 0; --j)
```
Fixed by 1faeb39.

philgekni commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+    return std::pair(std::ref(Stack[0]), i);
+}
+
+std::optional<std::reference_wrapper<MazePathfindingEntry>> MazePathfindingHistory::UpdateIntersectionEntry(
```
This name was to precise the R/W access to the entry. Anyway it looks that I don't use anymore this function and so cleaned it (71352ba).

philgekni commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4859,19 +4858,22 @@ void Guest::UpdateRideMazePathfinding()
 
     auto ride = get_ride(CurrentRide);
-    if (ride == nullptr)
+    if (!ride)
```
I updated the commit ca6c8fc in this way as well as make explicit another condition with a pointer at line 4894.

philgekni commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,132 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+    return Stack[0];
+}
+
+const std::array<MazePathfindingEntry, 5>& MazePathfindingHistory::ReadHistory() const
```
As you wish, fixed by 71352ba.

ZehMatt commented:
Well looking much better now, the bitfields still need to be addressed however.

ZehMatt commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4830,21 +4837,13 @@ void Guest::UpdateRideLeaveSpiralSlide()
+ */
+static constexpr uint16_t maze_forget_odds(double odds)
+{
+    return odds * UINT16_MAX;
```
I'm a bit worried that different compilers can produce different results here. This either needs a way to validate that the desired outcome is stored or better yet using integers only.

tupaschoal replied:
```suggestion
    return odds * std::numeric_limits<uint16_t>::max();
```

ZehMatt commented:
Sorry for the late response, I left one last comment otherwise looks good to me. Given that this is a rather larger change some of the other team members should also review this.

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
     }
 
     CoordsXY targetLoc = { DestinationX & 0xFFE0, DestinationY & 0xFFE0 };
+    uint16_t stationBaseZ = ride->stations[0].GetBaseZ();
```
```suggestion
    uint16_t stationBaseZ = static_cast<uint16_t>(ride->stations[0].GetBaseZ());
```

`GetBaseZ()` is an `int32_t`, implicitly casting to `uint16_t` might be a bad idea. Do you need to remove the signed though? (it was previously `int16_t`)

You only use it for the `map_get_track_element_at` which expects a `CoordsXYZ` object, so I'd say just call this directly there, without casting and such

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
     {
-        return;
-    }
+        TrackElement* trackElement = map_get_track_element_at({ targetLoc, stationBaseZ });
```
```suggestion
        auto trackElement = map_get_track_element_at({ targetLoc, stationBaseZ });
```

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
+        // If the subTile is connected to the maze entrance, close the edge giving to it
+        nearEntranceTile += CoordsDirectionDelta[entranceDirection];
+        entranceDirection = direction_reverse(entranceDirection);
+        if (nearEntranceTile.x == targetLoc.x && nearEntranceTile.y == targetLoc.y
```
```suggestion
        if (nearEntranceTile == targetLoc
```

This likely works

philgekni replied:
It doesn't look. I was tempted to get `nearEntranceTile` via `ToCoordsXY()` but I need the direction even if I don't care about the Z.

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
+        if (nearEntranceTile.x == targetLoc.x && nearEntranceTile.y == targetLoc.y
+            && (subTileIndex == entranceDirection || subTileIndex == direction_next(entranceDirection)))
+        {
+            mazeEntry |= 0x1 << gMazeCurrentDirectionToHedge[subTileIndex][direction_reverse(nearEntranceTile.direction)];
```
```suggestion
            mazeEntry |= 0x1 << gMazeCurrentDirectionToHedge[subTileIndex][entranceDirection];
```
`entranceDirection` is already `direction_reverse(nearEntranceTile.direction)`, is it not?

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
+        }
+
+        // Check open edges (without hedge)
+        for (int dir = 3; dir >= 0; --dir)
```
Do we need to iterate reversely? If not it would be great to use `ALL_DIRECTIONS` array,

philgekni replied:
I think so since I store the directions in `openEdgesFlags` with a shift. If needed I could rewrite this loop to use `ALL_DIRECTIONS`.

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
-            return;
+            if (currentEntry.IsCompletlyVisited())
+            {
+                // Offer amnesia to completly lost guest
```
```suggestion
                // Offer amnesia to completly lose guest
```
?

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4830,21 +4837,13 @@ void Guest::UpdateRideLeaveSpiralSlide()
+/**
+ * Helper function to define maze pathfinding thresholds
+ */
+static constexpr uint16_t maze_forget_odds(double odds)
```
```suggestion
static constexpr uint16_t MazeForgetOdds(double odds)
```

We're slowly converting everything to TitleCase, so better not adding any function in snake case.

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
+        };
+
+        // If the guest succeeded his memory roll, remove visited edges (other that the origin edge)
+        if ((scenario_rand() & 0xffff) >= rememberThreshold[entryIndex][peekFlag * 2])
```
```suggestion
        if ((scenario_rand() & 0xFFFF) >= rememberThreshold[entryIndex][peekFlag * 2])
```

tupaschoal commented on src/openrct2/peep/Guest.cpp:
```diff
@@ -4880,55 +4882,131 @@ void Guest::UpdateRideMazePathfinding()
+            openEdgesFlags &= ~currentEntry.GetVisited();
+
+        // Second dice roll, if succeeded remove also the origin edge (unless all the other edges are already visited)
+        if (bitcount(openEdgesFlags) > 1 && (scenario_rand() & 0xffff) >= rememberThreshold[entryIndex][1 + peekFlag * 2])
```
```suggestion
        if (bitcount(openEdgesFlags) > 1 && (scenario_rand() & 0xFFFF) >= rememberThreshold[entryIndex][1 + peekFlag * 2])
```

tupaschoal commented on src/openrct2/ride/gentle/Maze.h:
```diff
@@ -0,0 +1,178 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2019 OpenRCT2 developers
```
```suggestion
 * Copyright (c) 2014-2020 OpenRCT2 developers
```

tupaschoal commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,223 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+
+void MazePathfindingHistory::Initialize()
+{
+    for (auto& item : Stack)
```
```suggestion
    for (const auto& item : Stack)
```

philgekni replied:
I don't think that works as `item.Reset()` is non-const method.

tupaschoal commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,223 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+
+void MazePathfindingHistory::Free()
+{
+    for (auto& item : Stack)
```
```suggestion
    for (const auto& item : Stack)
```

philgekni replied:
Idem with `item.Free()`.

tupaschoal commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,223 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+std::pair<MazePathfindingEntry&, uint8_t> MazePathfindingHistory::MeetIntersection(
+    const CoordsXY& coords, uint8_t subTileIndex, Direction source)
+{
+    uint8_t i;
+
+    for (i = 0; i < Stack.size(); ++i)
+    {
+        if (Stack[i].MatchCoords(coords, subTileIndex))
+            break;
+    }
```
```suggestion
    auto matchingStack = std::find_if(std::begin(Stack), std::end(Stack), [&](const auto& entry){
        return entry.MatchCords(coords, subTileIndex);
    });
    auto i = matchingStack != std::end(Stack) ? std::distance(Stack.begin(), matchingStack) : 4;
```

I'm not convinced it's clearer, but I had already typed it anyway

Gymnasiast replied:
I think the `for` loop is a lot easier to read.

philgekni replied:
I agree with @Gymnasiast.

ZehMatt replied:
The purpose of std::find_if is to make it clear that this statement serves a single purpose and that is to find something, this also prevents users from adding up more code into the loop where it doesn't belong. std::find_if should be the preferred way in the C++ environment.

tupaschoal commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,223 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+            break;
+    }
+
+    for (uint8_t j = std::min((uint8_t)(Stack.size() - 1), i); j > 0; --j)
```
```suggestion
    for (uint8_t j = std::min(Stack.size() - 1, i); j > 0; --j)
```
I think you can let it implicitly cast safely

philgekni replied:
This is also my opinion but not msvc one.

tupaschoal commented on src/openrct2/ride/gentle/Maze.cpp:
```diff
@@ -181,3 +201,223 @@ TRACK_PAINT_FUNCTION get_track_paint_function_maze(int32_t trackType, int32_t di
+        item.Free();
+}
+
+std::pair<MazePathfindingEntry&, uint8_t> MazePathfindingHistory::MeetIntersection(
```
If this doesn't change the entry, it should probably return a const ref

philgekni replied:
In the Guest.c `UpdateRideMazePathfinding()` function, I modify the state of the returned entry via the method `MarkVisited`.

philgekni replied:
This is also my opinion but not msvc one.

Gymnasiast commented on src/openrct2-ui/windows/Guest.cpp:
```diff
@@ -2241,20 +2242,41 @@ void window_guest_debug_paint(rct_window* w, rct_drawpixelinfo* dpi)
+        {
+            auto [coords, subTileIndex] = maze_pathfind_item.GetCoords();
+            uint16_t args[] = {
+                (uint16_t)(coords.x >> 5),      (uint16_t)(coords.y >> 5),       subTileIndex,
```
That PR has now been merged.

ZehMatt commented:
@philgekni are you still working on this?

philgekni commented:
Hi @ZehMatt, I have unfortunately not the time currently to work on this task due to my work. I think that I'll be able to finish in March.

RobertWHurst commented:
So looking forward to this when you have time again. It's fantastic!

philgekni commented:
Thanks for the waiting, I rebased the branch and addressed your review comments.
I also fixed an error that I found in the pathfinding algorithm (b93363c).
Let me know if there is any problem, I should be more available now than by the past.

philgekni commented:
I continuing to work on the PR. I checked after record desynchronization.
I noticed that this kind of appears when a guest jumps into a maze. It is not at the frame where the function `Guest::UpdateRideMazePathfinding()` set `Action = PeepActionType::Jump;` but in the next frames during the animation.
I didn't yet narrow down the problem.

Dubaifreak commented:
The plugin can be useful, thanks for the download link!

karst commented:
Is this still planned on being finished?

rik-smeets commented:
@philgekni Do you still intend to work on this? If not, I will close this PR.

AaronVanGeffen commented:
The 'new' save format would allow for the maze pathfinding structs to be serialised into park saves, which was the original holdup. I hope this can be finished indeed.

philgekni commented:
Sorry I don't have anymore the time to work on this feature. Please feel free to close it or to take the ownership.

rik-smeets commented:
Thanks for responding, @philgekni.

I do agree it would be a neat feature, @AaronVanGeffen. Anyone wants to pick this one up? I don't want to keep this in limbo forever. Unfortunately, my limited C++ skillset won't allow me to do this one, so hopefully someone else wants to continue work on this.

tupaschoal commented:
I opened https://github.com/OpenRCT2/OpenRCT2/discussions/18599 to track this feature improvement and will close this PR for the time being, until someone can pick it up.
